File: services/product-api/app/telemetry/metrics.py
from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION, DEPLOYMENT_ENVIRONMENT
from opentelemetry.exporter.otlp.proto.grpc.metric_exporter import OTLPMetricExporter
import logging

logger = logging.getLogger(__name__)


class Metrics:
    """OpenTelemetry metrics for Product API."""
    
    def __init__(self):
        self.meter = None
        self.products_created_counter = None
        self.product_creation_duration = None
        self.active_requests_gauge = None
        self.kafka_publish_counter = None
        self.kafka_publish_errors = None
    
    def setup(self, service_name: str, service_version: str, environment: str, otlp_endpoint: str):
        """Initialize metrics with OTLP exporter."""
        resource = Resource(attributes={
            SERVICE_NAME: service_name,
            SERVICE_VERSION: service_version,
            DEPLOYMENT_ENVIRONMENT: environment,
            "service.namespace": "product-portfolio"
        })
        
        exporter = OTLPMetricExporter(
            endpoint=otlp_endpoint,
            insecure=True
        )
        
        reader = PeriodicExportingMetricReader(exporter, export_interval_millis=5000)
        
        provider = MeterProvider(resource=resource, metric_readers=[reader])
        metrics.set_meter_provider(provider)
        
        self.meter = metrics.get_meter(__name__)
        
        self.products_created_counter = self.meter.create_counter(
            name="products.created.total",
            description="Total number of products created",
            unit="1"
        )
        
        self.product_creation_duration = self.meter.create_histogram(
            name="products.creation.duration",
            description="Duration of product creation in seconds",
            unit="s"
        )
        
        self.active_requests_gauge = self.meter.create_up_down_counter(
            name="http.server.active_requests",
            description="Number of active HTTP requests",
            unit="1"
        )
        
        self.kafka_publish_counter = self.meter.create_counter(
            name="kafka.messages.published.total",
            description="Total number of messages published to Kafka",
            unit="1"
        )
        
        self.kafka_publish_errors = self.meter.create_counter(
            name="kafka.publish.errors.total",
            description="Total number of Kafka publish errors",
            unit="1"
        )
        
        logger.info("OpenTelemetry metrics initialized")


app_metrics = Metrics()

File: services/product-consumer/pkg/telemetry/metrics.go
package telemetry

import (
	"context"
	"log"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/exporters/otlp/otlpmetric/otlpmetricgrpc"
	"go.opentelemetry.io/otel/metric"
	sdkmetric "go.opentelemetry.io/otel/sdk/metric"
	"go.opentelemetry.io/otel/sdk/resource"
	semconv "go.opentelemetry.io/otel/semconv/v1.21.0"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"
)

type Metrics struct {
	MessagesConsumed   metric.Int64Counter
	ProcessingDuration metric.Float64Histogram
	MessagesInFlight   metric.Int64UpDownCounter
	ProcessingErrors   metric.Int64Counter
	DatabaseOperations metric.Int64Counter
}

var AppMetrics *Metrics

type MetricsConfig struct {
	ServiceName    string
	ServiceVersion string
	Environment    string
	OTLPEndpoint   string
}

func SetupMetrics(config MetricsConfig) (func(context.Context) error, error) {
	ctx := context.Background()

	conn, err := grpc.DialContext(ctx, config.OTLPEndpoint,
		grpc.WithTransportCredentials(insecure.NewCredentials()),
		grpc.WithBlock(),
	)
	if err != nil {
		return nil, err
	}

	exporter, err := otlpmetricgrpc.New(ctx, otlpmetricgrpc.WithGRPCConn(conn))
	if err != nil {
		return nil, err
	}

	res, err := resource.New(ctx,
		resource.WithAttributes(
			semconv.ServiceName(config.ServiceName),
			semconv.ServiceVersion(config.ServiceVersion),
			semconv.DeploymentEnvironment(config.Environment),
			attribute.String("service.namespace", "product-portfolio"),
		),
	)
	if err != nil {
		return nil, err
	}

	reader := sdkmetric.NewPeriodicReader(exporter)
	meterProvider := sdkmetric.NewMeterProvider(
		sdkmetric.WithResource(res),
		sdkmetric.WithReader(reader),
	)

	otel.SetMeterProvider(meterProvider)

	meter := meterProvider.Meter(config.ServiceName)

	AppMetrics = &Metrics{}

	AppMetrics.MessagesConsumed, err = meter.Int64Counter(
		"messages.consumed.total",
		metric.WithDescription("Total number of messages consumed from Kafka"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	AppMetrics.ProcessingDuration, err = meter.Float64Histogram(
		"messages.processing.duration",
		metric.WithDescription("Duration of message processing in seconds"),
		metric.WithUnit("s"),
	)
	if err != nil {
		return nil, err
	}

	AppMetrics.MessagesInFlight, err = meter.Int64UpDownCounter(
		"messages.in_flight",
		metric.WithDescription("Number of messages currently being processed"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	AppMetrics.ProcessingErrors, err = meter.Int64Counter(
		"messages.processing.errors.total",
		metric.WithDescription("Total number of message processing errors"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	AppMetrics.DatabaseOperations, err = meter.Int64Counter(
		"database.operations.total",
		metric.WithDescription("Total number of database operations"),
		metric.WithUnit("1"),
	)
	if err != nil {
		return nil, err
	}

	log.Printf("OpenTelemetry metrics initialized")

	return meterProvider.Shutdown, nil
}
